pkgs/webui.nix

```diff
@@ -2,157 +2,267 @@
+      substituteInPlace fetch_packages.py --replace \
+        "os.chdir(os.path.dirname(os.path.realpath(__file__)))" \
+        ""
 
-  webuiThirdPartyCache = stdenv.mkDerivation (webuiThirdPartyCommon // {
-    name = "contrail-webui-third-party-cache";
-    impureEnvVars = pkgs.stdenv.lib.fetchers.proxyImpureEnvVars;
-    outputHashMode = "recursive";
-    outputHashAlgo = "sha256";
-    outputHash = "1k5pmzx0gvpr015xy0463df9j38kzdmv74cpzx148m3is3zdd9w7";
-    postPatch = webuiThirdPartyCommon.postPatch + ''
+      # do not remove cached files
       substituteInPlace fetch_packages.py --replace \
         "os.remove(ccfile)" \
         "pass"
+
+      # do not ignore certificates when downloading
+      substituteInPlace fetch_packages.py --replace \
+        "subprocess.call(['wget', '--no-check-certificate', '-O', ccfile, url])" \
+        "subprocess.call(['curl', '-L', '-o', ccfile, url])"
     '';
-    buildPhase = "python fetch_packages.py -f packages.xml";
     installPhase = ''
-      mkdir -p $out/{cache,node_modules}
+      mkdir -p $out/bin
+      cp fetch_packages.py $out/bin/fetch_packages.py
+      chmod +x $out/bin/fetch_packages.py
+      wrapProgram $out/bin/fetch_packages.py  \
+        --prefix PATH : "${lib.makeBinPath [pythonWithXml unzip curl nodejs-4_x ]}" \
+        --set SSL_CERT_FILE ${cacert}/etc/ssl/certs/ca-bundle.crt
+    '';
+  };
 
-      # We remove some generated files and keep the npm cache
-      rm -rf node_modules/webworker-threads/
-      rm -rf cache/node_modules/webworker-threads/
-      cp -ra .npm $out/.npm/
+  webui-deps-rev = "${sources.webuiThirdParty.rev}";
+  webui-deps-file = "webui-deps-${webui-deps-rev}.tar.xz";
+  webui-deps-url = "https://storage.fr1.cloudwatt.com/v1/AUTH_e1cd9b90abb840798055d37b29f1a7d2/nixpkgs-tungsten-webui/${webui-deps-file}";
 
-      cp -ra cache/* $out/cache/
-      cp -ra node_modules/*.tar.gz $out/node_modules/
-    '';
-  });
+  # --- A NOTE on the Impurity of the webui dependencies and its workaround ---
+  #
+  # -- The Problem --
+  #
+  # Unfortunately the way in which the webui project fetches its dependencies is
+  # completely non-deterministic in its output. This is mostly because of 2 things:
+  #
+  # 1. ancient npm version
+  # 2. no package.json / lock file but independent `npm install` calls
+  #
+  # Updating npm and thus node is non-trivial and webui uses a lot of severly outdated packages
+  # and also applies patches that inject contrail specific code in dependencies.
+  #
+  # -- The workaround --
+  #
+  # Since the deps required to build the webui cannot be put into a derivation the shell
+  # expression below is provided to create a tarball. This tarball is then uploaded to an 
+  # object storage and is retrieved via `webui-thirdparty-deps` below.
+  #
+  # -- The workflow --
+  #
+  # If you need to update the webui using new sources you need to recreate the deps:
+  #
+  # $ nix-shell default.nix contrail50.webui.thirdPartyDeps
```

Comment:
```
$ nix-shell default.nix contrail50.webui.thirdPartyDeps
error: getting status of '/home/eon/vcs/nixpkgs-tungsten/contrail50.webui.thirdPartyDeps': No such file or directory
```

Reply:
There is a `-A` missing in my description... thanks for noticing!